docs/guides/ajax.rst

```diff
@@ -1,31 +1,220 @@
 Ajax
 ####
 
-Actions
-=======
+The ``elgg/Ajax`` AMD module (introduced in Elgg 2.1) provides a set of methods for communicating with the server in a concise and uniform way, which allows plugins to collaborate on the request data, the server response, and the returned client-side data.
 
-From JavaScript we can execute actions via XHR POST operations. Here's an example action and script for some basic math:
+.. contents:: Contents
+   :local:
+   :depth: 2
 
-.. code:: php
+Overview
+========
 
-    // in myplugin/actions/do_math.php
+All the ajax methods perform the following actions (not the hook types depend on the method):
 
-    if (!elgg_is_xhr()) {
-        register_error('Sorry, Ajax only!');
-        forward();
-    }
+#. Client-side, the ``data`` option is filtered by the hook ``ajax_request_data``.
+#. The request is made to the server, either rendering a view or a form, calling an action, or loading a path.
+#. Echoed JSON is turned into a response object (``Elgg\Services\AjaxResponse``).
+#. The response object is filtered by the hook ``ajax_response``.
+#. The response object is used to create the HTTP response.
+#. Client-side, the response data is filtered by the hook ``ajax_response_data``.
+#. The method returns a ``jqXHR`` object, which can be used as a Promise.
 
-    $arg1 = (int)get_input('arg1');
-    $arg2 = (int)get_input('arg2');
+More notes:
 
-    system_message('We did it!');
+* By default the ``elgg/spinner`` module is automatically used during requests.
+* User messages generated by ``system_message()`` and ``register_error()`` are collected and displayed on the client.
+* Elgg gives you a default error handler that shows a generic message if output fails.
+* PHP exceptions or denied resource return HTTP error codes, resulting in use of the client-side error handler.
+* The default HTTP method is POST for actions, otherwise GET. You can set it via ``options.method``.
+* For client caching, set ``options.method`` to "GET" and ``options.data.response_ttl`` to the max-age you want in seconds.
+* Do not use ``forward()`` with this API.
 
-    echo json_encode([
-        'sum' => $arg1 + $arg2,
-        'product' => $arg1 * $arg2,
-    ]);
+Performing actions
+------------------
+
+To execute an action, use ``performAction('action_name', options)``:
+
+.. code-block:: php
+
+	// in myplugin/actions/do_math.php
+
+	elgg_ajax_gatekeeper();
+
+	$arg1 = (int)get_input('arg1');
+	$arg2 = (int)get_input('arg2');
+
+	// will be rendered client-side
+	system_message('We did it!');
+
+	echo json_encode([
+		'sum' => $arg1 + $arg2,
+		'product' => $arg1 * $arg2,
+	]);
+
+.. code-block:: js
+
+	var ajax = new require('elgg/Ajax')();
+
+	ajax.performAction('do_math', {
+		data: {
+			arg1: 1,
+			arg2: 2
+		},
+	}).done(function (output) {
+		alert(output.sum);
+		alert(output.product);
+	});
+
+Notes for actions:
+
+* All hooks have type ``action:<action_name>``. In this case, "action:do_math".
```

Comment:
There will be 2 client-side hooks and 1 server-side, all with different hook names, all with this type. I guess I can make all those explicit in this section to get the point across.